README.md
# app-store-pricing-matrix

[![NPM version](https://img.shields.io/npm/v/app-store-pricing-matrix.svg?style=flat-square)](https://npmjs.org/package/app-store-pricing-matrix)
[![Build Status](https://img.shields.io/travis/amowu/app-store-pricing-matrix/master.svg?style=flat-square)](https://travis-ci.org/amowu/app-store-pricing-matrix) [![Coverage Status](https://img.shields.io/codecov/c/github/amowu/app-store-pricing-matrix/master.svg?style=flat-square)](https://codecov.io/gh/amowu/app-store-pricing-matrix/branch/master)

A simple module that holds currencies and prices from the Apple's iOS App Store.

## Install

npm:

    npm i app-store-pricing-matrix

Yarn:

    yarn add app-store-pricing-matrix

## Usage

```js
import AppStorePricingMatrix from "app-store-pricing-matrix";
```

or

```js
const AppStorePricingMatrix = require("app-store-pricing-matrix");
```

```js
const pricing = AppStorePricingMatrix.findBy({
  tier: "1",
  country: "US"
});

console.log(pricing.tierStem);        // => "1"
console.log(pricing.countryCode);     // => "US"
console.log(pricing.currencyCode);    // => "USD"
console.log(pricing.retailPrice);     // => 0.99
console.log(pricing.wholesalePrice);  // => 0.70
console.log(pricing.currencySymbol);  // => "$"
console.log(pricing.fRetailPrice);    // => "$ 0.99"
console.log(pricing.fWholesalePrice); // => "$ 0.70"
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [PricingInfo](#pricinginfo)
    -   [Properties](#properties)
-   [Tier](#tier)
    -   [Properties](#properties-1)
-   [FlattenedTier](#flattenedtier)
    -   [Properties](#properties-2)
-   [tiers](#tiers)
    -   [Examples](#examples)
-   [stems](#stems)
    -   [Examples](#examples-1)
-   [countries](#countries)
    -   [Examples](#examples-2)
-   [currencies](#currencies)
    -   [Examples](#examples-3)
-   [findBy](#findby)
    -   [Parameters](#parameters)
    -   [Examples](#examples-4)

### PricingInfo

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `countryCode` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The country code, e.g. `"US"`.
-   `currencyCode` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The currency code, e.g. `"USD"`.
-   `currencySymbol` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The currency symbol, e.g. `"$"`.
-   `fRetailPrice` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The retail price with currency symbol, e.g. `"$0.99"`.
-   `fWholesalePrice` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The wholesale price with currency symbol, e.g. `"$0.70"`.
-   `retailPrice` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The wholesale price, e.g. `0.99`.
-   `wholesalePrice` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The wholesale price, e.g. `0.70`.

### Tier

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `pricingInfo` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[PricingInfo](#pricinginfo)>** 
-   `tierName` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The name of tier, e.g. `"Tier 1"`.
-   `tierStem` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The stem of tier, e.g. `"1"`.

### FlattenedTier

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `countryCode` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The country code, e.g. `"US"`.
-   `currencyCode` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The currency code, e.g. `"USD"`.
-   `currencySymbol` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The currency symbol, e.g. `"$"`.
-   `fRetailPrice` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The retail price with currency symbol, e.g. `"$0.99"`.
-   `fWholesalePrice` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The wholesale price with currency symbol, e.g. `"$0.70"`.
-   `retailPrice` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The wholesale price, e.g. `0.99`.
-   `tierName` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The name of tier, e.g. `"Tier 1"`.
-   `tierStem` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The stem of tier, e.g. `"1"`.
-   `wholesalePrice` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The wholesale price, e.g. `0.70`.

### tiers

Get all items of tier.

#### Examples

```javascript
// returns
// [
//   {
//     "tierStem": "0",
//     "tierName": "Free",
//     "pricingInfo": [
//       {
//         "countryCode": "TW",
//         "currencyCode": "TWD",
//         "wholesalePrice": 0,
//         "retailPrice": 0,
//         "currencySymbol": "N",
//         "fRetailPrice": "NT$ 0",
//         "fWholesalePrice": "NT$ 0"
//       },
//       {
//         "countryCode": "US",
//         "currencyCode": "USD",
//         "wholesalePrice": 0.00,
//         "retailPrice": 0.00,
//         "currencySymbol": "$",
//         "fRetailPrice": "$0.00",
//         "fWholesalePrice": "$0.00"
//       },
//       {
//         "countryCode": "MY",
//         "currencyCode": "MYR",
//         "wholesalePrice": 0.00,
//         "retailPrice": 0.00,
//         "currencySymbol": "R",
//         "fRetailPrice": "RM0.00",
//         "fWholesalePrice": "RM0.00",
//       },
//       ...
//     ]
//   },
//   {
//     "tierStem": "1",
//     "tierName": "Tier 1",
//     "pricingInfo": [
//       {
//         "countryCode": "TW",
//         "currencyCode": "TWD",
//         "wholesalePrice": 20,
//         "retailPrice": 30,
//         "currencySymbol": "N",
//         "fRetailPrice": "NT$ 30",
//         "fWholesalePrice": "NT$ 20"
//       },
//       {
//         "countryCode": "US",
//         "currencyCode": "USD",
//         "wholesalePrice": 0.70,
//         "retailPrice": 0.99,
//         "currencySymbol": "$",
//         "fRetailPrice": "$0.99",
//         "fWholesalePrice": "$0.70"
//       },
//       {
//         "countryCode": "MY",
//         "currencyCode": "MYR",
//         "wholesalePrice": 2.73,
//         "retailPrice": 3.90,
//         "currencySymbol": "R",
//         "fRetailPrice": "RM3.90",
//         "fWholesalePrice": "RM2.73",
//       },
//       ...
//     ]
//   },
//   ...
//   {
//     "tierStem": "590",
//     "tierName": "Alternate Tier 5",
//     "pricingInfo": [
//       {
//         "countryCode": "TW",
//         "currencyCode": "TWD",
//         "wholesalePrice": 100,
//         "retailPrice": 150,
//         "currencySymbol": "N",
//         "fRetailPrice": "NT$ 150",
//         "fWholesalePrice": "NT$ 100"
//       },
//       {
//         "countryCode": "US",
//         "currencyCode": "USD",
//         "wholesalePrice": 3.50,
//         "retailPrice": 4.99,
//         "currencySymbol": "$",
//         "fRetailPrice": "$4.99",
//         "fWholesalePrice": "$3.50"
//       },
//       {
//         "countryCode": "MY",
//         "currencyCode": "MYR",
//         "wholesalePrice": 13.93,
//         "retailPrice": 19.90,
//         "currencySymbol": "R",
//         "fRetailPrice": "RM19.90",
//         "fWholesalePrice": "RM13.93",
//       },
//       ...
//     ]
//   }
// ]
AppStorePricingMatrix.tiers;
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Tier](#tier)>** An array

### stems

Get all stems of tier.

#### Examples

```javascript
// returns ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "40", "41", "42", "43", "44", "45", "46", "47", "48", "49", "50", "51", "52", "53", "54", "55", "56", "57", "58", "59", "60", "61", "62", "63", "64", "65", "66", "67", "68", "69", "70", "71", "72", "73", "74", "75", "76", "77", "78", "79", "80", "81", "82", "83", "84", "85", "86", "87", "510", "530", "550", "560", "570", "580", "590"]
AppStorePricingMatrix.stems;
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** An array of tier stems.

### countries

Get all country code of tier.

#### Examples

```javascript
// returns ["HK", "PT", "HN", "PY", "HR", "HU", "QA", "ID", "IE", "IL", "AE", "IN", "ZA", "IS", "AL", "IT", "AM", "AR", "AT", "AU", "RO", "RU", "BE", "BG", "JP", "BH", "BO", "SA", "BR", "SE", "SG", "SI", "BY", "SK", "CA", "SV", "CH", "KR", "CL", "CN", "CO", "CR", "KZ", "TH", "CY", "CZ", "TR", "DE", "TW", "TZ", "DK", "LT", "LU", "LV", "DO", "UA", "EC", "US", "EE", "EG", "MT", "MX", "MY", "ES", "VN", "NG", "NI", "NL", "NO", "FI", "NZ", "FR", "GB", "GR", "GT", "PA", "PE", "PH", "PK", "PL"]
AppStorePricingMatrix.countries
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** An array of country code.

### currencies

Get all currency code of tier.

#### Examples

```javascript
// returns ["HKD", "EUR", "USD", "HRK", "HUF", "QAR", "IDR", "ILS", "AED", "INR", "ZAR", "AUD", "RON", "RUB", "BGN", "JPY", "SAR", "BRL", "SEK", "SGD", "CAD", "CHF", "KRW", "CLP", "CNY", "COP", "KZT", "THB", "CZK", "TRY", "TWD", "TZS", "DKK", "EGP", "MXN", "MYR", "VND", "NGN", "NOK", "NZD", "GBP", "PEN", "PHP", "PKR", "PLN"]
AppStorePricingMatrix.currencies
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>** An array of currency code.

### findBy

Find pricing information by country code and tier stem.

#### Parameters

-   `pricing` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** 
    -   `pricing.country` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The country code, e.g. `"TW"`, `"US"`, `"MY"`.
    -   `pricing.tier` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The stem of tier.

#### Examples

"Tier 1" for USA


```javascript
// returns
// {
//   countryCode: "US",
//   currencyCode: "USD",
//   currencySymbol: "$",
//   fRetailPrice: "$0.99",
//   fWholesalePrice: "$0.7",
//   retailPrice: 0.99,
//   tierStem: "1",
//   wholesalePrice: 0.7
//  }
AppStorePricingMatrix.findBy({ country: "US", tier: "1" });
```

"Tier 1" for Taiwan


```javascript
// returns
// {
//   countryCode: "TW",
//   currencyCode: "TWD",
//   currencySymbol: "N",
//   fRetailPrice: "NT$ 30",
//   fWholesalePrice: "NT$ 20",
//   retailPrice: 30,
//   tierStem: "1",
//   wholesalePrice: 20
//  }
AppStorePricingMatrix.findBy({ country: "TW", tier: "1" });
```

"Tier 1" for Malaysia


```javascript
// returns
// {
//   countryCode: "MY",
//   currencyCode: "MYR",
//   currencySymbol: "R",
//   fRetailPrice: "RM3.90",
//   fWholesalePrice: "RM2.73",
//   retailPrice: 3.90,
//   tierStem: "1",
//   wholesalePrice: 2.73
//  }
AppStorePricingMatrix.findBy({ country: "MY", tier: "1" });
```

Returns **[FlattenedTier](#flattenedtier)** The pricing information.

## Updating the price

The pricing matrix JSON is taken from Apple's hidden API at the following URL.

    https://itunesconnect.apple.com/WebObjects/iTunesConnect.woa/ra/apps/#{AppStoreID}/pricing/matrix

The URL is accessible when you are logged in to Apple Connect with your browser.

When you need to update the pricing matrix,

1.  Login to Apple Connect with your developer account.
2.  Enter any App Store ID that you own.
3.  Your default browser will launch and load the JSON. Copy the entire content and paste into `input/pricing_matrix.json`.
4.  Run `npm test` to check if anything is broken.
5.  Send a pull request with updated `input/pricing_matrix.json`.

## Reference

Special thanks to @kenn [kenn/app_store_pricing_matrix](https://github.com/kenn/app_store_pricing_matrix)

## License

MIT © [Amo Wu](https://github.com/amowu)


CHANGELOG.md
## [1.0.6](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.5...v1.0.6) (2019-03-12)


### Bug Fixes

* fails to minify ([cb41501](https://github.com/amowu/app-store-pricing-matrix/commit/cb41501))



## [1.0.5](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.4...v1.0.5) (2019-03-07)


### Bug Fixes

* abort ES6 class ([1768cf6](https://github.com/amowu/app-store-pricing-matrix/commit/1768cf6))



## [1.0.4](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.3...v1.0.4) (2019-03-07)



## [1.0.3](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.2...v1.0.3) (2019-03-07)


### Bug Fixes

* findBy is not a function ([d077288](https://github.com/amowu/app-store-pricing-matrix/commit/d077288))



## [1.0.2](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.1...v1.0.2) (2019-03-07)


### Bug Fixes

* module not found error ([65b4dfd](https://github.com/amowu/app-store-pricing-matrix/commit/65b4dfd))



## [1.0.1](https://github.com/amowu/app-store-pricing-matrix/compare/v1.0.0...v1.0.1) (2019-03-07)



# 1.0.0 (2019-03-07)


### Features

* publish first version ([cba760e](https://github.com/amowu/app-store-pricing-matrix/commit/cba760e))






test/index.test.js
import { countries, currencies, findBy, stems } from "../src";

test("#stems", () => {
  const nonAlternateStems = Array.from(Array(88).keys()).map(String);
  const alternateStems = ["510", "530", "550", "560", "570", "580", "590"];

  const expectedStems = [...nonAlternateStems, ...alternateStems];

  expect(stems).toEqual(expectedStems);
});

test("#countries", () => {
  const expectedCountries = [
    "HK",
    "PT",
    "HN",
    "PY",
    "HR",
    "HU",
    "QA",
    "ID",
    "IE",
    "IL",
    "AE",
    "IN",
    "ZA",
    "IS",
    "AL",
    "IT",
    "AM",
    "AR",
    "AT",
    "AU",
    "RO",
    "RU",
    "BE",
    "BG",
    "JP",
    "BH",
    "BO",
    "SA",
    "BR",
    "SE",
    "SG",
    "SI",
    "BY",
    "SK",
    "CA",
    "SV",
    "CH",
    "KR",
    "CL",
    "CN",
    "CO",
    "CR",
    "KZ",
    "TH",
    "CY",
    "CZ",
    "TR",
    "DE",
    "TW",
    "TZ",
    "DK",
    "LT",
    "LU",
    "LV",
    "DO",
    "UA",
    "EC",
    "US",
    "EE",
    "EG",
    "MT",
    "MX",
    "MY",
    "ES",
    "VN",
    "NG",
    "NI",
    "NL",
    "NO",
    "FI",
    "NZ",
    "FR",
    "GB",
    "GR",
    "GT",
    "PA",
    "PE",
    "PH",
    "PK",
    "PL"
  ];

  expect(countries).toEqual(expectedCountries);
});

test("#currencies", () => {
  const expectedCurrencies = [
    "HKD",
    "EUR",
    "USD",
    "HRK",
    "HUF",
    "QAR",
    "IDR",
    "ILS",
    "AED",
    "INR",
    "ZAR",
    "AUD",
    "RON",
    "RUB",
    "BGN",
    "JPY",
    "SAR",
    "BRL",
    "SEK",
    "SGD",
    "CAD",
    "CHF",
    "KRW",
    "CLP",
    "CNY",
    "COP",
    "KZT",
    "THB",
    "CZK",
    "TRY",
    "TWD",
    "TZS",
    "DKK",
    "EGP",
    "MXN",
    "MYR",
    "VND",
    "NGN",
    "NOK",
    "NZD",
    "GBP",
    "PEN",
    "PHP",
    "PKR",
    "PLN"
  ];

  expect(currencies).toEqual(expectedCurrencies);
});

test("#findBy", () => {
  const tierUS0 = findBy({
    tier: "0",
    country: "US"
  });

  expect(tierUS0.tierStem).toEqual("0");
  expect(tierUS0.countryCode).toEqual("US");
  expect(tierUS0.currencyCode).toEqual("USD");
  expect(tierUS0.retailPrice).toEqual(0);
  expect(tierUS0.wholesalePrice).toEqual(0);

  const tierUS1 = findBy({
    tier: "1",
    country: "US"
  });

  expect(tierUS1.tierStem).toEqual("1");
  expect(tierUS1.countryCode).toEqual("US");
  expect(tierUS1.currencyCode).toEqual("USD");
  expect(tierUS1.retailPrice).toEqual(0.99);
  expect(tierUS1.wholesalePrice).toEqual(0.7);

  const tierTW1 = findBy({
    tier: "1",
    country: "TW"
  });

  expect(tierTW1.tierStem).toEqual("1");
  expect(tierTW1.countryCode).toEqual("TW");
  expect(tierTW1.currencyCode).toEqual("TWD");
  expect(tierTW1.retailPrice).toEqual(30);
  expect(tierTW1.wholesalePrice).toEqual(20);

  const tierMY1 = findBy({
    tier: "1",
    country: "MY"
  });

  expect(tierMY1.tierStem).toEqual("1");
  expect(tierMY1.countryCode).toEqual("MY");
  expect(tierMY1.currencyCode).toEqual("MYR");
  expect(tierMY1.retailPrice).toEqual(3.9);
  expect(tierMY1.wholesalePrice).toEqual(2.73);
});


src/index.js
/**
 * @typedef {Object} PricingInfo
 * @property {string} countryCode - The country code, e.g. `"US"`.
 * @property {string} currencyCode - The currency code, e.g. `"USD"`.
 * @property {string} currencySymbol - The currency symbol, e.g. `"$"`.
 * @property {string} fRetailPrice - The retail price with currency symbol, e.g. `"$0.99"`.
 * @property {string} fWholesalePrice - The wholesale price with currency symbol, e.g. `"$0.70"`.
 * @property {number} retailPrice - The wholesale price, e.g. `0.99`.
 * @property {number} wholesalePrice - The wholesale price, e.g. `0.70`.
 */

/**
 * @typedef {Object} Tier
 * @property {PricingInfo[]} pricingInfo
 * @property {string} tierName - The name of tier, e.g. `"Tier 1"`.
 * @property {string} tierStem - The stem of tier, e.g. `"1"`.
 */

/**
 * @typedef {Object} FlattenedTier
 * @property {string} countryCode - The country code, e.g. `"US"`.
 * @property {string} currencyCode - The currency code, e.g. `"USD"`.
 * @property {string} currencySymbol - The currency symbol, e.g. `"$"`.
 * @property {string} fRetailPrice - The retail price with currency symbol, e.g. `"$0.99"`.
 * @property {string} fWholesalePrice - The wholesale price with currency symbol, e.g. `"$0.70"`.
 * @property {number} retailPrice - The wholesale price, e.g. `0.99`.
 * @property {string} tierName - The name of tier, e.g. `"Tier 1"`.
 * @property {string} tierStem - The stem of tier, e.g. `"1"`.
 * @property {number} wholesalePrice - The wholesale price, e.g. `0.70`.
 */

import _, { find, get } from "lodash";

import jsonData from "../lib/input/pricing_matrix.json";

/**
 * Get all items of tier.
 * @static
 * @returns {Tier[]} An array
 * @example
 * // returns
 * // [
 * //   {
 * //     "tierStem": "0",
 * //     "tierName": "Free",
 * //     "pricingInfo": [
 * //       {
 * //         "countryCode": "TW",
 * //         "currencyCode": "TWD",
 * //         "wholesalePrice": 0,
 * //         "retailPrice": 0,
 * //         "currencySymbol": "N",
 * //         "fRetailPrice": "NT$ 0",
 * //         "fWholesalePrice": "NT$ 0"
 * //       },
 * //       {
 * //         "countryCode": "US",
 * //         "currencyCode": "USD",
 * //         "wholesalePrice": 0.00,
 * //         "retailPrice": 0.00,
 * //         "currencySymbol": "$",
 * //         "fRetailPrice": "$0.00",
 * //         "fWholesalePrice": "$0.00"
 * //       },
 * //       {
 * //         "countryCode": "MY",
 * //         "currencyCode": "MYR",
 * //         "wholesalePrice": 0.00,
 * //         "retailPrice": 0.00,
 * //         "currencySymbol": "R",
 * //         "fRetailPrice": "RM0.00",
 * //         "fWholesalePrice": "RM0.00",
 * //       },
 * //       ...
 * //     ]
 * //   },
 * //   {
 * //     "tierStem": "1",
 * //     "tierName": "Tier 1",
 * //     "pricingInfo": [
 * //       {
 * //         "countryCode": "TW",
 * //         "currencyCode": "TWD",
 * //         "wholesalePrice": 20,
 * //         "retailPrice": 30,
 * //         "currencySymbol": "N",
 * //         "fRetailPrice": "NT$ 30",
 * //         "fWholesalePrice": "NT$ 20"
 * //       },
 * //       {
 * //         "countryCode": "US",
 * //         "currencyCode": "USD",
 * //         "wholesalePrice": 0.70,
 * //         "retailPrice": 0.99,
 * //         "currencySymbol": "$",
 * //         "fRetailPrice": "$0.99",
 * //         "fWholesalePrice": "$0.70"
 * //       },
 * //       {
 * //         "countryCode": "MY",
 * //         "currencyCode": "MYR",
 * //         "wholesalePrice": 2.73,
 * //         "retailPrice": 3.90,
 * //         "currencySymbol": "R",
 * //         "fRetailPrice": "RM3.90",
 * //         "fWholesalePrice": "RM2.73",
 * //       },
 * //       ...
 * //     ]
 * //   },
 * //   ...
 * //   {
 * //     "tierStem": "590",
 * //     "tierName": "Alternate Tier 5",
 * //     "pricingInfo": [
 * //       {
 * //         "countryCode": "TW",
 * //         "currencyCode": "TWD",
 * //         "wholesalePrice": 100,
 * //         "retailPrice": 150,
 * //         "currencySymbol": "N",
 * //         "fRetailPrice": "NT$ 150",
 * //         "fWholesalePrice": "NT$ 100"
 * //       },
 * //       {
 * //         "countryCode": "US",
 * //         "currencyCode": "USD",
 * //         "wholesalePrice": 3.50,
 * //         "retailPrice": 4.99,
 * //         "currencySymbol": "$",
 * //         "fRetailPrice": "$4.99",
 * //         "fWholesalePrice": "$3.50"
 * //       },
 * //       {
 * //         "countryCode": "MY",
 * //         "currencyCode": "MYR",
 * //         "wholesalePrice": 13.93,
 * //         "retailPrice": 19.90,
 * //         "currencySymbol": "R",
 * //         "fRetailPrice": "RM19.90",
 * //         "fWholesalePrice": "RM13.93",
 * //       },
 * //       ...
 * //     ]
 * //   }
 * // ]
 * AppStorePricingMatrix.tiers;
 */
export const tiers = get(jsonData, ["data", "pricingTiers"], []);

/**
 * Get all stems of tier.
 * @static
 * @returns {string[]} An array of tier stems.
 * @example
 * // returns ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "40", "41", "42", "43", "44", "45", "46", "47", "48", "49", "50", "51", "52", "53", "54", "55", "56", "57", "58", "59", "60", "61", "62", "63", "64", "65", "66", "67", "68", "69", "70", "71", "72", "73", "74", "75", "76", "77", "78", "79", "80", "81", "82", "83", "84", "85", "86", "87", "510", "530", "550", "560", "570", "580", "590"]
 * AppStorePricingMatrix.stems;
 */
export const stems = tiers.map(({ tierStem }) => tierStem);

/**
 * Get all country code of tier.
 * @static
 * @returns {string[]} An array of country code.
 * @example
 * // returns ["HK", "PT", "HN", "PY", "HR", "HU", "QA", "ID", "IE", "IL", "AE", "IN", "ZA", "IS", "AL", "IT", "AM", "AR", "AT", "AU", "RO", "RU", "BE", "BG", "JP", "BH", "BO", "SA", "BR", "SE", "SG", "SI", "BY", "SK", "CA", "SV", "CH", "KR", "CL", "CN", "CO", "CR", "KZ", "TH", "CY", "CZ", "TR", "DE", "TW", "TZ", "DK", "LT", "LU", "LV", "DO", "UA", "EC", "US", "EE", "EG", "MT", "MX", "MY", "ES", "VN", "NG", "NI", "NL", "NO", "FI", "NZ", "FR", "GB", "GR", "GT", "PA", "PE", "PH", "PK", "PL"]
 * AppStorePricingMatrix.countries
 */
export const countries = _.chain(
  tiers.map(({ pricingInfo }) =>
    pricingInfo.map(({ countryCode }) => countryCode)
  )
)
  .flatten()
  .uniq()
  .value();

/**
 * Get all currency code of tier.
 * @static
 * @returns {string[]} An array of currency code.
 * @example
 * // returns ["HKD", "EUR", "USD", "HRK", "HUF", "QAR", "IDR", "ILS", "AED", "INR", "ZAR", "AUD", "RON", "RUB", "BGN", "JPY", "SAR", "BRL", "SEK", "SGD", "CAD", "CHF", "KRW", "CLP", "CNY", "COP", "KZT", "THB", "CZK", "TRY", "TWD", "TZS", "DKK", "EGP", "MXN", "MYR", "VND", "NGN", "NOK", "NZD", "GBP", "PEN", "PHP", "PKR", "PLN"]
 * AppStorePricingMatrix.currencies
 */
export const currencies = _.chain(
  tiers.map(({ pricingInfo }) =>
    pricingInfo.map(({ currencyCode }) => currencyCode)
  )
)
  .flatten()
  .uniq()
  .value();

/**
 * Find pricing information by country code and tier stem.
 * @static
 * @param {Object} pricing
 * @param {string} pricing.country - The country code, e.g. `"TW"`, `"US"`, `"MY"`.
 * @param {string} pricing.tier - The stem of tier.
 * @returns {FlattenedTier} The pricing information.
 * @example <caption>"Tier 1" for USA</caption>
 * // returns
 * // {
 * //   countryCode: "US",
 * //   currencyCode: "USD",
 * //   currencySymbol: "$",
 * //   fRetailPrice: "$0.99",
 * //   fWholesalePrice: "$0.7",
 * //   retailPrice: 0.99,
 * //   tierStem: "1",
 * //   wholesalePrice: 0.7
 * //  }
 * AppStorePricingMatrix.findBy({ country: "US", tier: "1" });
 * @example <caption>"Tier 1" for Taiwan</caption>
 * // returns
 * // {
 * //   countryCode: "TW",
 * //   currencyCode: "TWD",
 * //   currencySymbol: "N",
 * //   fRetailPrice: "NT$ 30",
 * //   fWholesalePrice: "NT$ 20",
 * //   retailPrice: 30,
 * //   tierStem: "1",
 * //   wholesalePrice: 20
 * //  }
 * AppStorePricingMatrix.findBy({ country: "TW", tier: "1" });
 * @example <caption>"Tier 1" for Malaysia</caption>
 * // returns
 * // {
 * //   countryCode: "MY",
 * //   currencyCode: "MYR",
 * //   currencySymbol: "R",
 * //   fRetailPrice: "RM3.90",
 * //   fWholesalePrice: "RM2.73",
 * //   retailPrice: 3.90,
 * //   tierStem: "1",
 * //   wholesalePrice: 2.73
 * //  }
 * AppStorePricingMatrix.findBy({ country: "MY", tier: "1" });
 */
export const findBy = ({ country, tier }) => {
  const { tierStem, tierName, pricingInfo } = find(tiers, {
    tierStem: tier
  });

  const pricingInfoFindByCountry = find(pricingInfo, {
    countryCode: country
  });

  return {
    tierStem,
    tierName,
    ...pricingInfoFindByCountry
  };
};
